src/Core/Indexes/KDTree.cs

```diff
@@ -0,0 +1,724 @@
+
+            /// <summary>
+            /// The left child of this node.
+            /// </summary>
+            private KDTreeNode leftChild;
+
+            /// <summary>
+            /// The right child of this node.
+            /// </summary>
+            private KDTreeNode rightChild;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="KDTreeNode"/> class.
+            /// </summary>
+            /// <param name="point">The point to be stored in this node.</param>
+            /// <param name="splitDimension">The splitting dimension of this node.</param>
+            /// <param name="totalDimensions">The total number of dimensions of the tree.</param>
+            public KDTreeNode(Coordinate point, int splitDimension, int totalDimensions)
+            {
+                this.point = point;
+                this.leftChild = null;
+                this.rightChild = null;
+                this.splitDimension = splitDimension;
+                this.TotalDimensions = totalDimensions;
+            }
+
+            /// <summary>
+            /// Gets the point stored in this node.
+            /// </summary>
+            public Coordinate Point
+            {
+                get
+                {
+                    return this.point;
+                }
+            }
+
+            /// <summary>
+            /// Gets or sets the total dimensions of the tree.
+            /// </summary>
+            public int TotalDimensions { get; set; }
+
+            /// <summary>
+            /// Gets a value indicating whether this node has a left child.
+            /// </summary>
+            public Boolean LeftChildExists { get { return this.leftChild != null; } }
+
+            /// <summary>
+            /// Gets a value indicating whether this node has a right child.
+            /// </summary>
+            public Boolean RightChildExists { get { return this.rightChild != null; } }
+
+            /// <summary>
+            /// Gets the number of geometries stored in this node and all its descendants.
+            /// </summary>
+            public int NumberOfGeometries
+            {
+                get
+                {
+                    int count = 0;
+
+                    if (this.LeftChildExists)
+                        count += this.leftChild.NumberOfGeometries;
+                    if (this.RightChildExists)
+                        count += this.rightChild.NumberOfGeometries;
+
+                    count += 1;
+                    return count;
+                }
+            }
+
+            /// <summary>
+            /// Adds a coordinate to the subtree of this node.
+            /// </summary>
+            /// <param name="coordinate">The coordinate to be added.</param>
+            public void Add(Coordinate coordinate)
+            {
+                if (coordinate.Equals(this.Point))
+                    throw new ArgumentException("Cannot add the same point twice.", "coordinate");
```

Comment:
Do you have a suggestion to which message should be used when attempting to perform a nearest neighbour search on an empty tree? I changed the rest as suggested.